File: mobile/androidApp/src/main/java/dev/screengoated/toolbox/mobile/ui/DonateSection.kt
package dev.screengoated.toolbox.mobile.ui

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.core.net.toUri
import dev.screengoated.toolbox.mobile.R
import dev.screengoated.toolbox.mobile.ui.i18n.MobileLocaleText

private const val DONATE_ACCOUNT = "8850273958"
private const val DONATE_BANK_LINE = "BIDV · NGUYEN BAO LINH · STK: 8850273958"
private const val VIETQR_URL =
    "https://img.vietqr.io/image/970418-8850273958-compact2.png" +
        "?accountName=NGUYEN%20BAO%20LINH&addInfo=Ung%20ho%20SGT"

// Collapsible donation card (collapsed by default), styled to match the other settings cards:
// a 40dp gradient-icon badge + title header (see ExpressiveSettingsHeader) with an expand
// chevron pinned right. The Vietnamese bank transfer (VietQR) is shown in every language;
// EN/KO add a note that it's for Vietnamese donors only. The QR is referenced by its VietQR
// image URL (opened externally), never bundled as an asset.
@Composable
internal fun DonateSection(
    locale: MobileLocaleText,
    modifier: Modifier = Modifier,
) {
    val context = LocalContext.current
    val accent = MaterialTheme.colorScheme.primary
    var expanded by rememberSaveable { mutableStateOf(false) }

    ExpressiveSettingsCard(accent = accent, modifier = modifier.fillMaxWidth()) {
        Column(verticalArrangement = Arrangement.spacedBy(ShellSpacing.itemGap)) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .clickable { expanded = !expanded },
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(ShellSpacing.itemGap),
            ) {
                Box(
                    modifier = Modifier
                        .size(40.dp)
                        .background(accent.copy(alpha = 0.18f), MaterialTheme.shapes.medium),
                    contentAlignment = Alignment.Center,
                ) {
                    GradientMaskedIcon(
                        iconRes = R.drawable.ms_potted_plant,
                        brush = Brush.linearGradient(
                            listOf(accent, MaterialTheme.colorScheme.primary),
                        ),
                        modifier = Modifier.size(20.dp),
                    )
                }
                Text(
                    text = locale.donateHeader,
                    style = MaterialTheme.typography.titleMedium,
                    modifier = Modifier.weight(1f),
                )
                Icon(
                    painter = painterResource(
                        if (expanded) R.drawable.ms_expand_less else R.drawable.ms_expand_more,
                    ),
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.size(24.dp),
                )
            }

            if (expanded) {
                Text(
                    text = locale.donateBody,
                    style = MaterialTheme.typography.bodyMedium,
                )
                Text(
                    text = DONATE_BANK_LINE,
                    style = MaterialTheme.typography.bodyMedium,
                    fontWeight = FontWeight.SemiBold,
                    color = accent,
                )
                Row(horizontalArrangement = Arrangement.spacedBy(ShellSpacing.itemGap)) {
                    ExpressiveSettingsButton(
                        text = "Sao chép STK",
                        onClick = {
                            val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE)
                                as ClipboardManager
                            clipboard.setPrimaryClip(ClipData.newPlainText("STK", DONATE_ACCOUNT))
                        },
                        accent = accent,
                    )
                    ExpressiveSettingsButton(
                        text = "Mở mã VietQR",
                        onClick = {
                            runCatching {
                                context.startActivity(
                                    Intent(Intent.ACTION_VIEW, VIETQR_URL.toUri())
                                        .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK),
                                )
                            }
                        },
                        accent = accent,
                    )
                }
                // EN/KO get an extra line clarifying the bank transfer is for Vietnamese donors.
                if (!locale.donateVietnamese) {
                    Text(
                        text = locale.donateNote,
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
            }
        }
    }
}

File: src/gui/settings_ui/global/donate_section.rs
//! Donation section shown in global settings.
//!
//! Donations currently support Vietnamese donors only. For the Vietnamese locale the
//! section shows the VietQR bank-transfer details + actions; English/Korean locales show
//! a short "Vietnamese donors only" note instead. The QR is referenced by its VietQR
//! image URL (opened in the browser) rather than bundled as an asset.
//!
//! The section is collapsed by default — a `potted_plant` glyph + title on the left, an
//! expand chevron pinned to the right; the whole header row toggles it.

use crate::gui::icons::{self, Icon};
use crate::gui::locale::LocaleText;
use eframe::egui;

const DONATE_ACCOUNT: &str = "8850273958";
const DONATE_BANK_LINE: &str = "BIDV · NGUYEN BAO LINH · STK: 8850273958";
const VIETQR_URL: &str = "https://img.vietqr.io/image/970418-8850273958-compact2.png?accountName=NGUYEN%20BAO%20LINH&addInfo=Ung%20ho%20SGT";

pub fn render_donate_section_content(ui: &mut egui::Ui, text: &LocaleText) {
    let id = ui.make_persistent_id("sgt_donate_collapsing");
    let mut state =
        egui::collapsing_header::CollapsingState::load_with_default_open(ui.ctx(), id, false);
    let openness = state.openness(ui.ctx());

    let header_resp = ui
        .horizontal(|ui| {
            icons::draw_icon_static(ui, Icon::PottedPlant, Some(icons::ICON_MD));
            ui.label(
                egui::RichText::new(text.global_settings.donate_header)
                    .strong()
                    .size(14.0),
            );
            ui.with_layout(egui::Layout::right_to_left(egui::Align::Center), |ui| {
                let size = ui.spacing().icon_width;
                let (rect, _) =
                    ui.allocate_exact_size(egui::vec2(size, size), egui::Sense::hover());
                let chevron = if openness < 0.5 {
                    Icon::ArrowDown
                } else {
                    Icon::ArrowUp
                };
                icons::paint_icon(
                    ui.painter(),
                    rect,
                    chevron,
                    ui.visuals().widgets.inactive.fg_stroke.color,
                );
            });
        })
        .response
        .interact(egui::Sense::click());

    if header_resp.clicked() {
        state.toggle(ui);
    }
    if header_resp.hovered() {
        ui.ctx().set_cursor_icon(egui::CursorIcon::PointingHand);
    }
    state.store(ui.ctx());

    state.show_body_unindented(ui, |ui| {
        ui.add_space(6.0);
        ui.label(text.global_settings.donate_body);
        ui.add_space(4.0);
        ui.label(egui::RichText::new(DONATE_BANK_LINE).strong());
        ui.add_space(6.0);
        ui.horizontal(|ui| {
            if ui.button("Sao chép STK").clicked() {
                ui.ctx().copy_text(DONATE_ACCOUNT.to_string());
            }
            if ui.button("Mở mã VietQR").clicked() {
                let _ = open::that(VIETQR_URL);
            }
        });
        // EN/KO get an extra line clarifying the bank transfer is for Vietnamese donors.
        if !text.global_settings.donate_vietnamese {
            ui.add_space(6.0);
            ui.label(
                egui::RichText::new(text.global_settings.donate_note)
                    .size(11.5)
                    .color(crate::gui::theme::AppTheme::from_ui(ui).on_surface_variant()),
            );
        }
    });
}
